hazelcast-client/src/main/java/com/hazelcast/client/HazelcastClientManager.java

```diff
@@ -0,0 +1,156 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.client;
+
+import java.util.HashSet;
+import java.util.Collection;
+import java.util.Collections;
+
+import com.hazelcast.util.EmptyStatement;
+
+import java.util.concurrent.ConcurrentMap;
+
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.core.OutOfMemoryHandler;
+
+import java.util.concurrent.ConcurrentHashMap;
+
+import com.hazelcast.client.config.ClientConfig;
+import com.hazelcast.client.impl.HazelcastClientProxy;
+import com.hazelcast.instance.OutOfMemoryErrorDispatcher;
+import com.hazelcast.core.DuplicateInstanceNameException;
+import com.hazelcast.client.config.XmlClientConfigBuilder;
+import com.hazelcast.client.impl.HazelcastClientInstanceImpl;
+import com.hazelcast.client.impl.ClientConnectionManagerFactory;
+import com.hazelcast.client.impl.DefaultClientConnectionManagerFactory;
+
+public class HazelcastClientManager {
+    /***
+     * Instance for clientManagers
+     */
+    public static final HazelcastClientManager HAZELCAST_CLIENT_MANAGER = new HazelcastClientManager();
+
+    static {
+        OutOfMemoryErrorDispatcher.setClientHandler(new ClientOutOfMemoryHandler());
+    }
+
+    private static final ConcurrentMap<String, HazelcastClientProxy> CLIENTS
+            = new ConcurrentHashMap<String, HazelcastClientProxy>(5);
+
+    public HazelcastInstance newHazelcastClient(HazelcastClientFactory hazelcastClientFactory) {
+        return newHazelcastClient(new XmlClientConfigBuilder().build(), hazelcastClientFactory);
+    }
+
+    public HazelcastInstance newHazelcastClient(ClientConfig config, HazelcastClientFactory hazelcastClientFactory) {
+        if (config == null) {
+            config = new XmlClientConfigBuilder().build();
+        }
+
+        final ClassLoader tccl = Thread.currentThread().getContextClassLoader();
+        HazelcastClientProxy proxy;
+        try {
+            Thread.currentThread().setContextClassLoader(HazelcastClient.class.getClassLoader());
+            ClientConnectionManagerFactory clientConnectionManagerFactory = new DefaultClientConnectionManagerFactory();
+            final HazelcastClientInstanceImpl client = hazelcastClientFactory.createHazelcastInstanceClient(
+                    config,
+                    clientConnectionManagerFactory
+            );
+            client.start();
+            OutOfMemoryErrorDispatcher.registerClient(client);
+            proxy = hazelcastClientFactory.createProxy(client);
+
+            if (CLIENTS.containsKey(client.getName())) {
```

Comment:
Could it be that we get some racy behavior since the check/modify is not atomic?